Rename EFFECT_FAKE_OUT to EFFECT_FIRST_TURN_ONLY

AlexOn1ine opened the pull request:
Renamed `EFFECT_FAKE_OUT` to `EFFECT_FIRST_TURN_ONLY` to more reflect it's use. Also changed AI logic slightly to differentiate between fake out and first impression.

AsparagusEduardo commented:
Hang on, mechanically, what are the main differences between the two? I was sure that First Impression was a more powerful Fake Out.

AlexOn1ine commented:
> Hang on, mechanically, what are the main differences between the two? I was sure that First Impression was a more powerful Fake Out.

First Impression doesn't flinch

AsparagusEduardo commented:
> First Impression doesn't flinch
Then wouldn't that fall under `additionalEffects` for Fake Out?

AlexOn1ine commented:
> > First Impression doesn't flinch
> > Then wouldn't that fall under `additionalEffects` for Fake Out?

No, not necessarily. 100% flinching is a strong effect which imo deserves it's own case. Otherwise it would never be chosen because of weak power. Similarly First Impression has 90 BP and should if possible be preferred over other moves with a similar power.

AsparagusEduardo commented:
> No, not necessarily. 100% flinching is a strong effect which imo deserves it's own case. Otherwise it would never be chosen because of weak power. Similarly First Impression has 90 BP and should if possible be preferred over other moves with a similar power and because it has higher priority.

Ah, this is for AI.

I disagree completely, the AI should be adjusted instead so that it considers the 100% flinch instead of fragmenting the data into more effects.

AlexOn1ine commented:
> > No, not necessarily. 100% flinching is a strong effect which imo deserves it's own case. Otherwise it would never be chosen because of weak power. Similarly First Impression has 90 BP and should if possible be preferred over other moves with a similar power and because it has higher priority.
> 
> Ah, this is for AI.
> 
> I disagree completely, the AI should be adjusted instead so that it considers the 100% flinch instead of fragmenting the data into more effects.

Without some major rewrite this isn't possible. Fake Out and First Impression need an effect because they always fail from the 2nd turn on wards. Additionally Fake out has a Priority of 3. This sets it's apart from other flinch effect moves. Also moves with 100% chance to flinch are already handled in `AI_CalcMoveScore`.

cfmnephrite commented:
I think this could do with a rethink. I suspect the reason you're doing this isn't because they have separate effects but rather because it makes AI calculation easier which is not a sustainable strategy. With additional effects, they unambiguously have the same effect i.e. "fail if not first turn".

Hence, both moves need to run a lot of the same checks in `ShouldFakeOut` - the problem is that Fake Out has an additional check for Inner Focus but this should no longer be the case. Both moves should have the exact same check for first turn or not, then the additional check for any additional effects. Otherwise you're taking a step back by giving moves their own unique effects again.

AlexOn1ine commented:
> Hence, both moves need to run a lot of the same checks in `ShouldFakeOut` 

No, they have different checks from each other. First Impression should never check `ShouldFakeOut` because non of the conditions apply there. 

I disagree that it's a step back. Some move effects are unique and need to be distinguished from each other. A lot of those moves have negative side effects which we somehow need to signal to the AI

AlexOn1ine commented:
> > Hence, both moves need to run a lot of the same checks in `ShouldFakeOut`
> 
> No, they have different checks from each other. First Impression should never check `ShouldFakeOut` because non of the conditions apply there.
> 
> I disagree that it's a step back. Some move effects are unique and need to be distinguished from each other. A lot of those moves have negative side effects which we somehow need to signal to the AI

I'm open for suggestion but we either have to keep the hardcode check for First Impression or make a new effect for it.

AlexOn1ine commented:
Changed `EFFECT_FAKE_OUT` to `EFFECT_FIRST_TURN_ONLY`, removed the new first impression effect to keep using the now new named effect and added an argument to differentiate both moves in `AI_CalcMoveScore`.

cfmnephrite commented on src/battle_ai_main.c:
```diff
@@ -3875,14 +3877,11 @@ static u32 AI_CalcMoveScore(u32 battlerAtk, u32 battlerDef, u32 move)
-        }
+    case EFFECT_FIRST_TURN_ONLY:
+        if (gBattleMoves[move].argument == MOVE_FAKE_OUT && ShouldFakeOut(battlerAtk, battlerDef, move))
+            ADJUST_SCORE(4);
```
Instead of checking the move argument, wouldn't simply checking the following be enough?
```
HasMoveEffectWithChance(MOVE_EFFECT_FLINCH, 100)```

cfmnephrite replied:
I think you're right in that Fake Out needs special consideration, btw, sorry for not being more helpful. I think its unique combination of priority, effect and additional effects makes it tricky to catch - but I think the above suggestion is the most sustainable, no?

AlexOn1ine replied:
> but I think the above suggestion is the most sustainable, no?
ye, it looks like it. I can remove both arguments now.

AlexOn1ine replied:
pushed a change

AlexOn1ine commented:
> I think its unique combination of priority, effect and additional effects makes

Main concern is that we need to signal the AI that those moves fail on the second turn. If it wasn't for that we would be able to handle it without the effect.

AsparagusEduardo commented on src/data/battle_move_effects.h:
```diff
@@ -811,7 +811,7 @@ const struct BattleMoveEffect gBattleMoveEffects[NUM_BATTLE_MOVE_EFFECTS] =
-    [EFFECT_FAKE_OUT] =
+    [EFFECT_FIRST_TURN_ONLY] =
     {
         .battleScript = BattleScript_EffectFakeOut,
```
It needs to rename the script as well

AlexOn1ine replied:
fixed

AsparagusEduardo reviewed (changes requested):
Can you please add the other changes you did to the PR description since it's not just Fake Out/First Impression? Besides that and the script rename, lgtm.

AlexOn1ine commented:
> Can you please add the other changes you did to the PR description since it's not just Fake Out/First Impression? Besides that and the script rename, lgtm.

changed

AsparagusEduardo commented on src/data/battle_move_effects.h:
```diff
@@ -1114,6 +1108,12 @@ const struct BattleMoveEffect gBattleMoveEffects[NUM_BATTLE_MOVE_EFFECTS] =
         .encourageEncore = TRUE,
     },
 
+    [EFFECT_PLACEHOLDER] =
+    {
+        .battleScript = BattleScript_EffectPlaceholder,
+        .battleTvScore = 0, // TODO: Assign points
+    },
+
```
Why move this back?

AlexOn1ine replied:
Idk how this happened. Seeing this change for the first time. I looked through the commit history and it looks like the change appeared after I merged upcoming. I'll fix it tomorrow. 

> I was about to say approve, but found one more thing, sorry D:

all good

AlexOn1ine replied:
If you want you can do a suggested change and I'll accept it

AlexOn1ine replied:
Fixed

AsparagusEduardo reviewed (changes requested):
I was about to say approve, but found one more thing, sorry D:

AsparagusEduardo commented:
Approved, though about the `remove useful comment`, I removed it because we're never going to assign TV points for to the placeholder effect ^^;
